File: internal/schema/shard.go
/*
 * paperback: resilient paper backups for the very paranoid
 * Copyright (C) 2018 Aleksa Sarai <cyphar@cyphar.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package schema

import (
	"bytes"
	"encoding/json"
	"strings"

	"github.com/cyphar/paperback/internal/version"
	"github.com/cyphar/paperback/pkg/crypto"
	"github.com/cyphar/paperback/pkg/shamir"
)

// Shard is a wrapped version of shamir.Share, containing the version data for
// the current schema. We don't bother making a "deep" clone of the structure
// here, because we control the shamir package and so won't accidentally break
// the schema. The encrypted variant can be created using Encrypt().
type Shard struct {
	Version string       `json:"version"`
	Inner   shamir.Share `json:"share"`
}

// NewShard constructs a new schema.Shard from a shamir.Share, using the
// default values for all other fields. This is the recommended way of creating
// a new Shard.
func NewShard(share shamir.Share) Shard {
	return Shard{
		Version: version.Version,
		Inner:   share,
	}
}

// EncryptedShard is the type of the armored and OpenPGP encrypted blob that is
// published as the shard key payload. It can be decrypted using Decrypt. This
// is just an alias for []byte but is used for type-safety.
type EncryptedShard []byte

// Encrypt returns the encrypted version of the given shard (using the provided
// key) of type EncryptedShard. This is the recommended way of handling
// encryption of shards as this way it will remain consistent.
func (s Shard) Encrypt(key []byte) (EncryptedShard, error) {
	// Construct the buffer with the plaintext shard.
	shardBytes, err := json.Marshal(s)
	if err != nil {
		return nil, err
	}
	shardBuffer := bytes.NewBuffer(shardBytes)

	// Encrypt the shard.
	encryptedShard, err := crypto.Encrypt(shardBuffer, key)
	if err != nil {
		return nil, err
	}
	return encryptedShard, nil
}

// Decrypt returns the decrypted version (or an error) of the given shard using
// the given key. This is the recommended way of handling decryption of shards
// as this way all users will remain consistent.
func (es EncryptedShard) Decrypt(key []byte) (Shard, error) {
	// Decrypt the shard.
	encryptedShardBuffer := bytes.NewBuffer(es)
	shardBytes, err := crypto.Decrypt(encryptedShardBuffer, key)
	if err != nil {
		return Shard{}, err
	}

	// Unmarshal the JSON.
	var shard Shard
	if err := json.Unmarshal(shardBytes, &shard); err != nil {
		return Shard{}, err
	}
	// TODO: Probably should do some sort of crypto verification here, to
	//       ensure that the shard is actually useful.
	return shard, nil
}

File: pkg/paperback-core/src/shamir/shard.rs
/*
 * paperback: paper backup generator suitable for long-term storage
 * Copyright (C) 2018-2022 Aleksa Sarai <cyphar@cyphar.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

use crate::{
    shamir::gf::{GfElem, GfElemPrimitive},
    v0::{FromWire, ShardId, ToWire},
};

use unsigned_varint::{encode as varuint_encode, nom as varuint_nom};

/// Piece of a secret which has been sharded with [Shamir Secret Sharing][sss].
///
/// [sss]: https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Shard {
    pub(super) x: GfElem,
    pub(super) ys: Vec<GfElem>,
    pub(super) secret_len: usize,
    pub(super) threshold: GfElemPrimitive,
}

impl Shard {
    pub const ID_LENGTH: usize = 8;

    /// Returns the *unique* identifier for a given `Shard`.
    ///
    /// If two shards have the same identifier, they cannot be used together for
    /// secret recovery.
    pub fn id(&self) -> ShardId {
        multibase::encode(multibase::Base::Base32Z, &self.x.to_bytes())
    }

    /// Returns the number of *unique* sister `Shard`s required to recover the
    /// stored secret.
    pub fn threshold(&self) -> u32 {
        self.threshold
    }
}

pub fn parse_id(id: ShardId) -> Result<GfElem, multibase::Error> {
    let (_, data) = multibase::decode(id)?;
    Ok(GfElem::from_bytes(data))
}

impl ToWire for Shard {
    fn to_wire(&self) -> Vec<u8> {
        let mut bytes = vec![];

        // Encode x-value.
        varuint_encode::u32(self.x.inner(), &mut varuint_encode::u32_buffer())
            .iter()
            .for_each(|b| bytes.push(*b));

        // Encode y-values (length-prefixed).
        varuint_encode::usize(self.ys.len(), &mut varuint_encode::usize_buffer())
            .iter()
            .copied()
            .chain(self.ys.iter().flat_map(|y| {
                varuint_encode::u32(y.inner(), &mut varuint_encode::u32_buffer()).to_owned()
            }))
            .for_each(|b| bytes.push(b));

        // Encode threshold.
        varuint_encode::u32(self.threshold, &mut varuint_encode::u32_buffer())
            .iter()
            .for_each(|b| bytes.push(*b));

        // Encode secret length.
        varuint_encode::usize(self.secret_len, &mut varuint_encode::usize_buffer())
            .iter()
            .for_each(|b| bytes.push(*b));

        bytes
    }
}

impl FromWire for Shard {
    fn from_wire_partial(input: &[u8]) -> Result<(&[u8], Self), String> {
        use nom::{combinator::complete, multi::many_m_n, IResult};

        fn parse(input: &[u8]) -> IResult<&[u8], Shard> {
            let (input, x) = varuint_nom::u32(input)?;
            let x = GfElem::from_inner(x);

            let (input, ys_length) = varuint_nom::usize(input)?;
            let (input, ys) = many_m_n(ys_length, ys_length, varuint_nom::u32)(input)?;
            let ys = ys
                .iter()
                .copied()
                .map(GfElem::from_inner)
                .collect::<Vec<_>>();

            let (input, threshold) = varuint_nom::u32(input)?;
            let (input, secret_len) = varuint_nom::usize(input)?;

            Ok((
                input,
                Shard {
                    x,
                    ys,
                    secret_len,
                    threshold,
                },
            ))
        }
        let mut parse = complete(parse);

        let (input, shard) = parse(input).map_err(|err| format!("{:?}", err))?;

        Ok((input, shard))
    }
}

#[cfg(test)]
impl quickcheck::Arbitrary for Shard {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        Self {
            x: GfElem::arbitrary(g),
            ys: (0..g.size()).map(|_| GfElem::arbitrary(g)).collect(),
            secret_len: usize::arbitrary(g),
            threshold: u32::arbitrary(g),
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[quickcheck]
    fn shard_bytes_roundtrip(shard: Shard) {
        let shard2 = Shard::from_wire(&shard.to_wire()).unwrap();
        assert_eq!(shard, shard2);
    }
}
